MAINT:stats:Cythonize and remove Fortran statlib code

ilayn opened the pull request:
#### Reference issue
Part of #18566 

#### What does this implement/fix?

I converted the private FORTRAN code that was living under `statslib`. Though not so long it was quite tricky to follow all the integer tricks. Then I cythonized them for a reduced performance hit. 

The code is very old (from 1980s and added to SciPy 20+years ago) and only works with that particular style. I can sense that this can be so much more performant by regular numpy usage however I have no idea what the code does hence followed the old code closely and tested as much as I can. 

#### Additional information
 - swilk was returning a for no apparent reason. Now `a` is modified in place and not returned. Also everything now works in double precision and not the weird `real/float` mix.
 - I can guarantee output of `gscale` within a meshgrid of (100x100). It starts overflowing around (50x50) anyways but tested regardless Also tested `m+n = 150` cases with `m=1, 2, 10, 20`. 
 - For `swilk` tested many conditions including nan inputs and other cases also passed the tests on my machine with no modification. The decimal digits are preserved  to 12-14 digits. 
 - `prho` was never used and no tests failed with its absence so directly removed it.

ilayn commented:
> This looks like swilk: https://real-statistics.com/tests-normality-and-symmetry/statistical-tests-normality-symmetry/shapiro-wilk-expanded-test/

I appreciate your trust in my stats knowledge but it is really not up to par to code something in SciPy hence I'll leave it to the stats residents.

mdhaber commented:
I've implemented Shapiro-Wilk in Python, but I don't have complaints with the Fortran that I wanted to address at this time.

ilayn commented on scipy/stats/meson.build:
```diff
@@ -14,35 +14,17 @@ stats_special_cython_gen = generator(cython,
-  fortran_args: fortran_ignore_warnings,
-  link_with: statlib_lib,
-  dependencies: [fortranobject_dep],
+_stats = py3.extension_module('_stats',
```
Not sure why GH is showing the diff as such but I basically removed only the statlib parts.

mdhaber commented:
@ilayn is Cython vs Pythran a personal preference here, or does Cython have an important advantage here?

ilayn commented:
I didn't think about Pythran at all. After conversion Python version was slow so I bumped the speed up. 

The impact is not much since now both are around microseconds, for `gscale` 760ns vs 1.75 microseconds and so on. Unless someone runs this on a 10000 for loop negligible difference.

mdhaber commented:
In that case, may I suggest using Pythran? Cython is more approachable than Fortran, but Pythran is even better from that perspective. gh-18328 comes to mind - if `swilk` had been Python-like instead of Fortran, we probably would have applied the fix in `swilk` itself rather than adding that special case in `shapiro`.

ilayn commented:
I don't think there is significant difference. Cython is also Python code just verbose type definitions at the top of the functions. You can literally copy paste your fix to this function here without any impact. Cython compiles as much as it can with explicit defs and leaves it like regular python code hence I don't think Pythran or Cython matters. 

https://github.com/scipy/scipy/pull/18679/files#diff-e395b1c082d3122404921ba93c6b06aa20b6b4e841d8b426f86dbd34bc7c7e4cR328-R330

h-vetinari commented:
> I don't think there is significant difference.

It depends on the example in question of course, but there can be _significant_ differences, like factor 3-5 faster. Of course that's not guaranteed to always happen, but the difference to Cythons is kinda Pythran's main raison d'être. Pythran also has more higher-level constructs (also allows using more of numpy's API in the pythran code than cython), and seems to get optimize better under compilation.

mdhaber commented:
Oh, I assumed Cython was faster. If Pythran is faster, that's icing on the cake. I was after the maintainability aspect, and I was comfortable even with a speed hit. I've authored PRs with both, and since the early build issues were resolved (I'm on Windows, too), Pythran has been much easier to work with.

ilayn commented:
I mean if it was that nice i would have switched too. But there's no such free lunch, feel free to copy the body and convert to Pythran, there is no cython command it's all python code all copy pasteable. So I don't know what you mean by maintainability aspect.

For the rest it's all C and Cython code already in stats hence I'm not sure exactly where the resistance coming from.

mdhaber commented:
It's not resistance, really. It's just the only opinion I have here. Feel free to ignore.

ilayn commented:
No I mean I'm interested too if there is a significant gain of course but I want to know if there is any issues cython is causing. I just don't want to write double code because currently Pyhran is opt-in and not always enabled. 

> It depends on the example in question of course, but there can be significant differences,

Yes but there is no way to find out when as far as I know (which is very little). Cython has the annotation to tell you where the Python interaction is hence that's what I mean by consistency.

h-vetinari commented:
No resistance from my side either, just wanted to clarify.

> I just don't want to write double code because currently Pyhran is opt-in and not always enabled.

100% agreed about not writing twice. OTOH, pythran's been opt-in for a long time (since 1.7), and https://github.com/scipy/scipy/issues/13308 was closed recently. I think we could retake a look at making it mandatory, or at a timeline for that (e.g. start warning if pythran isn't present and see if someone complains). CC @rgommers

ilayn commented:
Ah finally CI checks out, this one is good to go from my side.

dschmitz89 commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,463 @@
+    if n > 5000:
+        ifault = 2
+
+    SA /= n1
+    SX /= n1
+    SSA, SSX, SAX = 0., 0., 0.
+    ind2 = n - 1
+    for ind1 in range(n1):
+        if ind1 != ind2:
+            ASA = (-1 if ind1 < ind2 else 1)*a[min(ind1, ind2)] - SA
+        else:
+            ASA = -SA
+
+        XSX = x[ind1]/RANGE - SX
+        SSA += ASA * ASA
+        SSX += XSX * XSX
+        SAX += ASA * XSX
+        ind2 -= 1
+
+    SSASSX = sqrt(SSA * SSX)
```
I hate to be be "that guy" but if we keep the original FORTRAN style, this will still be hard to maintain in my opinion. Are all capitalized variables here matrices or was that convention just taken from the original?

I am not a statistical testing guy though, so cannot judge how complex this would be to implement in plain numpy or readable cython so.

ilayn replied:
All coming from the fortran code. I just wanted to give whoever is reading this a chance to compare with the fortran file. If I knew a bit more about the algorithm, I'd give it a go for a numpy-ized version but this is just to get out of the fortran burden. 

I think a stats person would recognize the steps here better and can modify things in place. Because I really don't get what is going on here. But on the upside, other than variable naming, it's all arithmetics thanks to being written in the olden days, where there is nothing available.

ilayn replied:
By the way, the `mvndst` code, that I am attempting with pythran, is even worse to decipher which includes a manual cholesky decomposition and pivoting. So fingers crossed :)

mdhaber replied:
@ilayn don't worry about mvndst. It is basically done in gh-17410; it just hasn't been made public.

ilayn replied:
Ah you mean, we don't need to convert it ?

ilayn replied:
Indeed that looks like the case. I think I can get rid of that permuted cholesky part with a regular lapack machinery if you want. Ah I can't be happier that I don't need to go over the remaining code to be honest heheh.

mdhaber replied:
>  I think I can get rid of that permuted cholesky part with a regular lapack machinery if you want.

Probably. Here are a few thoughts. 

Eliminating Fortran code, especially code without any bug reports, is not yet a high priority to me. (Personally, I was hoping that we would replace our code with `git submodule`s maintained by the modern Fortran folks. Part of the problem with maintaining them is the language, but I think there are resource limitations, too, so it would be good to offload some of that work since there are others interested in doing it. Anyway, that explains why I am not very enthusiastic about some of these Fortran-replacement PRs.)

That said, pythranizing `_mvt.py` is something I wanted to do anyway. It would close an issue filed by a user (gh-16142) and make `multivariate_t` much faster. So I would be willing to review a PR that pythranizes `_mvt.py` and replaces use of `mvndst`. I would suggest that it start with only essential changes to minimize the diff. The main change I anticipate is that `rng` inputs need to be replaced with a pre-generated random array of the appropriate size (unless Pythran has added the ability to accept NumPy RNGs as input). I'd ask that code to be pythranized be left in `_qmvnt.py` and any code that is not to be pythranized be added to a new file or `multivariate.py`. 

After that is done, replacing the permuted Cholesky with LAPACK would be a welcome follow-up PR.

ilayn replied:
I used to believe the same thing; why change if things are working without any issues. But as you mentioned a few of them, these are 50 year old code that is practically unreadable. And it's based on a dying tech (F77 not implying anything about newer Fortran). Hence after reading all these fortran codes for the last month, I can safely say that argument about stability won't fly anymore. These code need to be get rid of. They are stopping us to even test ideas: copy paste some code, try it out, and so on. Even LAPACK folks are questioning the future of fortran let alone these individual code. 

I'm skipping the actual issues with F77 which we have beaten up to death elsewhere. But in particular with mnv I meant only the particular cholesky part that I know what it is doing. The remaining parts, I really don't have any capacity to integrate.

mdhaber replied:
> The remaining parts, I really don't have any capacity to integrate.

Right. If I seemed to ask for something else, let me know what it was so I can clarify.

rgommers replied:
> Personally, I was hoping that we would replace our code with `git submodule`s maintained by the modern Fortran folks.

I have so far seen some stated enthusiasm, but very few (if any) contributions on SciPy itself or SciPy-equivalent code. Outside of working on LFortran itself being able to compiler SciPy of course. So aside from the question of what's most desirable - that depends on code complexity I'd say - I don't see this happening any time soon.

The stats/special routines are overall pretty simple compared to large components like sparse solver libraries, and I don't think they are candidates for separate Fortran packages. So I'd say a 1:1 translation of them to Python/Cython/Pythran is a clear win.

mdhaber replied:
> The stats/special routines are overall pretty simple compared to large components like sparse solver libraries, and I don't think they are candidates for separate Fortran packages. So I'd say a 1:1 translation of them to Python/Cython/Pythran is a clear win

Yes, I shouldn't have brought up the modern Fortran stuff, since I wouldn't expect them to take over this simple stuff.

I agree that it's a win if no bugs are introduced and we can find a reviewer for it.

ilayn replied:
At least for this PR, all tests I've done are double precision identical. The reason I'm confident about that is because I also compiled the file externally and run the tests against each other. 

So if there are no `if` branches I've missed, I'd say this is pretty much identical.

mdhaber replied:
Ok. I can't review the code, but I can write some tests shortly.

mdhaber commented:
I tried comparing the two version of `gscale`. When I run the following code with the new version:

```python3
from scipy.stats._AB_SW import gscale as gscale2
astarts_pr = []
a1s_pr = []
for i in range(10):
    for j in range(10):
        print(i, j)
        astart_pr, a1_pr, _ = gscale2(i, j)
        astarts_pr.append(astart_pr)
        a1s_pr.append(a1_pr)  # comment out this line

```
The program just stops after about 10 iterations (both on command line and in IDE. In IDE, it says `Process finished with exit code -1073740940 (0xC0000374)`). If I comment out the last line, it runs to completion.

Note that I'm still doing an in-place build with `setup.py`, in case that could have something to do with it.

ilayn commented:
That's probably, I forgot to guard against i, j being 0. If you start with `range(1,10)` it would probably go through

mdhaber commented:
Weird... yes, it does - but why? `gscale2(0, 0)` works, and there is not a problem unless I append the a1 result to the list.

In any case, that case would never be called by `stats.ansari`, and otherwise the output of the new `gscale` does seems to be identical to that of the original. 

<details>

```python3
import numpy as np
from scipy.stats._statlib import gscale
from scipy.stats._AB_SW import gscale as gscale2
from time import perf_counter

time_main = 0
time_pr = 0
for i in range(1, 200):
    for j in range(1, 200):
        print(i, j)

        tic = perf_counter()
        astart_main, a1_main, _ = gscale(i, j)
        toc = perf_counter()
        time_main += toc - tic

        tic = perf_counter()
        astart_pr, a1_pr, _ = gscale2(i, j)
        toc = perf_counter()
        time_pr += toc - tic

        np.testing.assert_equal(a1_pr, a1_main)
        assert astart_pr==astart_main

print(time_pr/time_main)
```

</details>

It's very slightly slower on my machine, which is fine IMO. 

I can confirm "I can guarantee output of gscale within a meshgrid of (100x100). It starts overflowing around (50x50) anyways but tested regardless". I actually went up to 200.

I could suggest that if the `dtype` were changed to 64-bit float, we could go up higher without overflow, but there is probably no reason to, and whether there is or not is a separate issue. So I have nothing against the `gscale` part of this PR. On to `swilk`.

mdhaber commented:
For `swilk`, the agreement is not as close. It's may just be roundoff error being amplified, but we try to be careful about these things in `stats`. Consider an extreme case:

```python3
import numpy as np
from scipy.stats._AB_SW import swilk
y = np.sort([-180089.49350026142, -180089.49185713136, -180089.48736901287, -180089.48231540626,  -180089.46989864993])
n = len(y)
a = np.zeros(n//2, dtype=np.float64)
init = 0
_, w, pw, ifault = swilk(y.copy(), a.copy(), init)
w  # now 0.8985956472033496, was 0.2524259686470032
pw  # now 0.40217241177567253, was 3.542755065005787e-12
```

It has something to do with the variance being small relative to the location of the data, but any ideas about why the new code would produce a different result?

Here's the code I was using to stress-test this.

<details>

```python3
import numpy as np
from scipy import stats
from scipy.stats._statlib import swilk
from scipy.stats._AB_SW import swilk as swilk2
from time import perf_counter
rng = np.random.default_rng(9245923596356935)
time_main = 0
time_pr = 0
for n in range(4, 5000):
    for j in range(1, 3):
        print(n, j)

        a = np.zeros(n//2, dtype=np.float64)
        init = 0
        x = rng.normal(loc=stats.norm(scale=1e5).rvs(random_state=rng),
                       scale=stats.loguniform(1e-3, 1e6).rvs(random_state=rng),
                       size=n)
        y = np.sort(x)

        tic = perf_counter()
        _, w, pw, ifault = swilk(y.copy(), a.copy(), init)
        toc = perf_counter()
        time_main += toc - tic

        tic = perf_counter()
        w2, pw2, ifault2 = swilk2(y, a, init)
        toc = perf_counter()
        time_pr += toc - tic

        np.testing.assert_allclose(w2, w, rtol=1e-3)
        np.testing.assert_allclose(pw2, pw, rtol=1e-3)
        assert np.all(ifault2 == ifault)

print(time_pr/time_main)
```

</details>

ilayn commented:
For gscale; 
I think I know why, it's OK with `0` inputs. I would say `(0, 1)` is the offending input. At the end of `gscale` there is a condition that says 

```python
        # reverse array
        if other > test:
            for ind in range(LL//2):
                a1v[LL-1-ind], a1v[ind] = a1v[ind], a1v[LL-1-ind
```

when encountered with `(0, 1)` probably it is trying to revert it but because `boundscheck` is False secretly segfaulting for reaching out to an element outside memoryview. Then probably returning garbage that can't be converted to a python object. Then things I don't understand follow and resulting with a low-level problem. Seems like a trivial fix

For `swilk`, 

that's way more than I ever encountered and clearly unacceptable so thanks for that example. Probably it goes into some other branch. I'll check what is going on with the Fortran comparison.

ilayn commented:
For `swilk`, the differences are coming from working with float64. This algorithm doesn't give matching results with high precision and I don't know which one is the correct. In fact I can see how they slowly diverge with the additional precision. Is there any way we can test this behavior with some other tool?

ilayn commented:
Indeed converting to float32 makes the results match ends with 

```
4998 1
4998 2
4999 1
4999 2
2.542812896175186
```

ilayn commented:
Looking at the values again, did you take out the median from the test cases like `shapiro` does or it doesn't have any significance?

mdhaber commented:
It shouldn't affect the calculation, and it doesn't seem to (at least when floating point doesn't get in the way).
```python3
import numpy as np
from scipy import stats
stats.shapiro([1, 2, 3, 4])
stats.shapiro([10, 20, 30, 40])
stats.shapiro([15, 25, 35, 45])
# should all be the same
```

mdhaber commented:
We added the line to subtract the median in https://github.com/scipy/scipy/pull/15777 because the Fortran version had issues when the variance is relatively small compared to the data.

It turns out that this new version was more accurate than the Fortran in this case. (I suppose we should have known that the Fortran version was wrong the probability of getting a p-value of `3.542755065005787e-12` in one of the iterations above is .000000000354%.) After subtracting the median, we get approximately the same value with the new code and Fortran code, and this is approximately the same as the new code gets without subtracting the median.

Still, if you run the stress test above (corrected with subtraction of the median) you will see some discrepancies.

mdhaber commented on scipy/stats/_morestats.py:
```diff
@@ -1861,19 +1861,19 @@ def shapiro(x):
     if N < 3:
         raise ValueError("Data must be at least length 3.")
 
     x = x - np.median(x)
```
We can delete this and subtract from `y` below to save memory and an unnecessary `median` calculation.

ilayn replied:
Done.

mdhaber commented on scipy/stats/_morestats.py:
```diff
@@ -1861,19 +1861,19 @@ def shapiro(x):
+    a = zeros(N//2, dtype=np.float64)
     init = 0
 
     y = sort(x)
```
Fine to modify `y` in place since it's a copy.
```suggestion
    y = sort(x)
    y -= x[N//2]  # subtract the median (or a nearby value); see gh-15777
```

ilayn replied:
Done.

mdhaber commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,463 @@
+    w = 1 - w1
+
+    # Calculate significance level for W (exact for N=3)
+    if n == 3:
+        pw = PI6 * (asin(sqrt(w)) - PI_OVER_3)
+        return w, pw, ifault
```
In the Fortran version, this could result in negative p-values, so lines 1883-1886 of `_morestats.py` overrides the p-value for `n == 3`. If the Cython version here is more accurate, it's possible that those Python lines can be deleted. Alternatively, the formulation of the calculation in 1883-1886 can be moved here. (It's more obvious that the p-value cannot be negative in that formulation.)

ilayn replied:
I checked the offending input in the reported issue and Cython version also returns negative numbers. Must be something algorithmic hence I adopted your fix directly baked into the Cython version with a note about the modification.

mdhaber reviewed (commented):
I just have two suggestions for the `swilk` part of this PR, but neither is necessary.

ilayn commented:
I found the offending part in the gscale tests which was not immediately returning after small sample size. Now it is passing without any strange segfaults. 

For the swilk I added the modifications you suggested. Also added some DOIs for quick access to things as many as I could find. 

Currently, on my machine `gscale_pr`/`gscale_sp` is 1.33 and `swilk_pr`/`swilk_sp` is 2.13 (some slowdown is because of double precision anyways) which is I think acceptable.

ilayn commented:
Very nice. So apparently p value can return almost 0 and with some noise it can be marked as -eps. Do you have any suggestions how to handle that ? 

The remaining failures (except lint is not related).

mdhaber commented:
> So apparently p value can return almost 0 and with some noise it can be marked as -eps

If that's for $n=3$, we can generate an approximation that is valid where the output is nearly zero. Otherwise, the car we can do is `np.clip` to keep it non-negative.

ilayn commented:
Yes it is indeed. https://github.com/scipy/scipy/actions/runs/5369718336/jobs/9741361808?pr=18679#step:3:3190

The violation is very little -1.9526481126463935e-16 and only on Linux 32 bit runs which are lately turning out to be more problematic than we'd wish for it.

mdhaber commented:
Actually the problem is that the statistic is falling below 3/4. Maybe clip the statistic to keep it above 3/4 (theoretical minimum, apparently) before passing it into the formula for the p-value.

ilayn commented:
OK clamped the lower bound to 3/4. Let's see how it goes.

ilayn commented:
After a flaky TestDGamma failure this looks like passing now

mdhaber reviewed (approved):
The output of the new `gscale` and `swilk` looked good in my tests. I'll let someone else merge, since I am not the right person to review the Cython.

ilayn commented:
Thanks @mdhaber for the review anyways, since this is for 1.12, there is no rush for it, let's wait until someone else has bandwidth to check.

rgommers reviewed (commented):
Thanks for dealing with that code @ilayn. I went through the Cython code and reviewed for potential issues (not algorithmic correctness, I trust the tests and other reviewers for that).

ilayn commented on scipy/stats/_morestats.py:
```diff
@@ -13,7 +13,7 @@
 from scipy._lib._util import _rename_parameter, _contains_nan, _get_nan
 
-from . import _statlib
+from ._AB_SW import gscale, swilk
```
no preference and sure no problem :) `ansari-bradley-shapiro-wilk` seemed too long back when I started. Now used the fortran names `_ansari_swilk_statistics`

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+# cython: language_level=3
```
old habits

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    cdef cnp.ndarray a3 = cnp.PyArray_ZEROS(1, [LL], cnp.NPY_FLOAT32, 0)
+    cdef float[::1] a1v = a1  # memview of a1
+    cdef float[::1] a2v = a2  # memview of a2
+    cdef float[::1] a3v = a3  # memview of a3
```
Memoryiews are faster with element access and many for loops are done over views. But they can't do numpy ops like `@` or `+=`-like infix ops. So we create the numpy arrays for that kind of ops and put memoryview on top of their memory to do array access. It's really not very nice for my taste, but performance goes up somehow with memoryviews I guess via skipping NumPy overhead. 

Cython 3 is said to fix some of these things but can't say anything before trying.

rgommers replied:
Okay, if it helped let's keep it. It'd be useful if you pushed a commit with a comment to explain why the code is like this - because it's the kind of thing that the next person who comes along may try to clean up otherwise.

ilayn replied:
Yes let's do it when it's ready I'll add a skip ci comment commit

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    Null Distribution of the Ansari-Bradley W Statistic", Applied Statistics,
+    25(1), 1976, :doi:`10.2307/2346534`
+    """
+    cdef int m = min(test, other), n = max(test, other)
```
Done but it goes a bit out of control down the file so I kept the ones that assume a default value on separate and left the others on single line declarations

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    """
+    cdef int odd = n % 2
+    cdef float A = 1., B = 3., C = 2. if odd else 0.
+    cdef int ndo = (n + 2 + odd) // 2 - odd
```
added

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    return temp if upper else (1-temp)
+
+
+cdef double _ppnd(double p) noexcept:
```
We use  `abs` function in the body which is gil-requiring. Cython 3 fixed those declarations. But this function doesn't do much anyways hence probably more cost to release and lock GIL

rgommers replied:
ugh. guess I'll resolve it then, not much to be done about it now.

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    """
+    cdef double A0, A1, A2, A3, B1, B2, B3, B4, C0, C1, C2, C3, D1, D2
+    cdef double q, r, split = 0.42, temp
+    # cdef int ifault = 0
```
That's for someone who wants to enable ifault later at some point, I basically disabled it since we didn't use. But left it as a comment at the top uncomment if needed.

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+        r = q * q
+        temp = q*(((A3*r + A2)*r + A1) * r + A0)
+        temp = temp / ((((B4*r + B3)*r + B2)*r + B1)*r + 1.)
+        return temp  #, 0
```
added

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    """
+    cdef int m = min(test, other), n = max(test, other)
+    cdef int astart = ((test + 1) // 2) * (1 + (test // 2))
+    cdef int LL = (test * other) // 2 + 1
```
Unfortunately, since this is an exact test, it starts overflowing very quickly around `m=n=150`. Hence it won't reach critical  values to consider that.

rgommers replied:
Okay, sounds good. Resolving.

ilayn commented on scipy/stats/_AB_SW.pyx:
```diff
@@ -0,0 +1,501 @@
+    return nextlenb
+
+
+def swilk(double[::1] x, double[::1] a, bint init=False, int n1=-1):
```
It's a `def` func so it doesn't accept `nogil`. Even if it did, then we have to release it from Python which is not cost efficient in a single call like this.

ilayn commented:
Ah dammit, the rename broke the review comments. But thanks Ralf for the review. I think I addressed them all. Have a look when you have the time please.

ilayn commented:
OK comment added. It's ready to merge at your convenience.

rgommers commented:
In it went. Thanks @ilayn and @mdhaber & other reviewers.
